# Sales Data Assignment — Python/pandas Version

This notebook converts the original Google Sheets/Excel assignment into Python exercises and provides complete solutions.

**Dataset:** 25 sales records and 8 columns. The data intentionally contains a duplicate `OrderID` (`O1024`) so duplicate detection can be practised.

## Learning objectives
- Create and inspect a pandas DataFrame
- Filter and sort data
- Calculate totals, averages, counts, conditional summaries and percentages
- Detect duplicates
- Perform lookup operations
- Create pivot-style summaries
- Apply conditional styling
- Protect the original data by working on copies


## Setup and DataFrame creation

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: f'{x:,.2f}')

### Create the DataFrame directly from the assignment data

In [2]:
sales_data = [{'OrderID': 'O1001', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 5, 'UnitPrice': 500, 'Revenue': 2500}, {'OrderID': 'O1002', 'Salesperson': 'Ravi', 'Region': 'South', 'Product': 'Mouse', 'Category': 'Accessories', 'Units': 20, 'UnitPrice': 20, 'Revenue': 400}, {'OrderID': 'O1003', 'Salesperson': 'Neha', 'Region': 'West', 'Product': 'Keyboard', 'Category': 'Accessories', 'Units': 15, 'UnitPrice': 30, 'Revenue': 450}, {'OrderID': 'O1004', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 7, 'UnitPrice': 180, 'Revenue': 1260}, {'OrderID': 'O1005', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Printer', 'Category': 'Office', 'Units': 3, 'UnitPrice': 220, 'Revenue': 660}, {'OrderID': 'O1006', 'Salesperson': 'Ravi', 'Region': 'West', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 2, 'UnitPrice': 500, 'Revenue': 1000}, {'OrderID': 'O1007', 'Salesperson': 'Neha', 'Region': 'South', 'Product': 'Chair', 'Category': 'Furniture', 'Units': 6, 'UnitPrice': 90, 'Revenue': 540}, {'OrderID': 'O1008', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Desk', 'Category': 'Furniture', 'Units': 2, 'UnitPrice': 200, 'Revenue': 400}, {'OrderID': 'O1009', 'Salesperson': 'Asha', 'Region': 'West', 'Product': 'Mouse', 'Category': 'Accessories', 'Units': 25, 'UnitPrice': 20, 'Revenue': 500}, {'OrderID': 'O1010', 'Salesperson': 'Ravi', 'Region': 'North', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 4, 'UnitPrice': 180, 'Revenue': 720}, {'OrderID': 'O1011', 'Salesperson': 'Neha', 'Region': 'East', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 3, 'UnitPrice': 500, 'Revenue': 1500}, {'OrderID': 'O1012', 'Salesperson': 'Amit', 'Region': 'South', 'Product': 'Printer', 'Category': 'Office', 'Units': 4, 'UnitPrice': 220, 'Revenue': 880}, {'OrderID': 'O1013', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Keyboard', 'Category': 'Accessories', 'Units': 10, 'UnitPrice': 30, 'Revenue': 300}, {'OrderID': 'O1014', 'Salesperson': 'Ravi', 'Region': 'East', 'Product': 'Desk', 'Category': 'Furniture', 'Units': 1, 'UnitPrice': 200, 'Revenue': 200}, {'OrderID': 'O1015', 'Salesperson': 'Neha', 'Region': 'West', 'Product': 'Chair', 'Category': 'Furniture', 'Units': 8, 'UnitPrice': 90, 'Revenue': 720}, {'OrderID': 'O1016', 'Salesperson': 'Amit', 'Region': 'North', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 6, 'UnitPrice': 500, 'Revenue': 3000}, {'OrderID': 'O1017', 'Salesperson': 'Asha', 'Region': 'South', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 5, 'UnitPrice': 180, 'Revenue': 900}, {'OrderID': 'O1018', 'Salesperson': 'Ravi', 'Region': 'West', 'Product': 'Printer', 'Category': 'Office', 'Units': 2, 'UnitPrice': 220, 'Revenue': 440}, {'OrderID': 'O1019', 'Salesperson': 'Neha', 'Region': 'East', 'Product': 'Mouse', 'Category': 'Accessories', 'Units': 18, 'UnitPrice': 20, 'Revenue': 360}, {'OrderID': 'O1020', 'Salesperson': 'Amit', 'Region': 'South', 'Product': 'Keyboard', 'Category': 'Accessories', 'Units': 14, 'UnitPrice': 30, 'Revenue': 420}, {'OrderID': 'O1021', 'Salesperson': 'Asha', 'Region': 'North', 'Product': 'Desk', 'Category': 'Furniture', 'Units': 2, 'UnitPrice': 200, 'Revenue': 400}, {'OrderID': 'O1022', 'Salesperson': 'Ravi', 'Region': 'West', 'Product': 'Chair', 'Category': 'Furniture', 'Units': 5, 'UnitPrice': 90, 'Revenue': 450}, {'OrderID': 'O1023', 'Salesperson': 'Neha', 'Region': 'South', 'Product': 'Laptop', 'Category': 'Electronics', 'Units': 4, 'UnitPrice': 500, 'Revenue': 2000}, {'OrderID': 'O1024', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 9, 'UnitPrice': 180, 'Revenue': 1620}, {'OrderID': 'O1024', 'Salesperson': 'Amit', 'Region': 'East', 'Product': 'Monitor', 'Category': 'Electronics', 'Units': 9, 'UnitPrice': 180, 'Revenue': 1620}]
df = pd.DataFrame(sales_data)
df

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500,2500
1,O1002,Ravi,South,Mouse,Accessories,20,20,400
2,O1003,Neha,West,Keyboard,Accessories,15,30,450
3,O1004,Amit,East,Monitor,Electronics,7,180,1260
4,O1005,Asha,North,Printer,Office,3,220,660
5,O1006,Ravi,West,Laptop,Electronics,2,500,1000
6,O1007,Neha,South,Chair,Furniture,6,90,540
7,O1008,Amit,East,Desk,Furniture,2,200,400
8,O1009,Asha,West,Mouse,Accessories,25,20,500
9,O1010,Ravi,North,Monitor,Electronics,4,180,720


### Optional: read the same data directly from the supplied workbook

In [3]:
file_path = Path('PIIT_SalesData.csv')
df1 = pd.read_csv(file_path)
df1.head()

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500,2500
1,O1002,Ravi,South,Mouse,Accessories,20,20,400
2,O1003,Neha,West,Keyboard,Accessories,15,30,450
3,O1004,Amit,East,Monitor,Electronics,7,180,1260
4,O1005,Asha,North,Printer,Office,3,220,660


## 1. Freeze the top row

In pandas there is no scrolling worksheet to freeze. The Python equivalent is to keep column names visible while displaying data and, during Excel export, use `freeze_panes=(1, 0)`. Here we verify the headers.

In [4]:
df.columns.tolist()

['OrderID',
 'Salesperson',
 'Region',
 'Product',
 'Category',
 'Units',
 'UnitPrice',
 'Revenue']

## 2. Apply a filter and sort Revenue in descending order

Filter-like operations use Boolean indexing; sorting uses `sort_values()`.

In [5]:
sorted_by_revenue = df.sort_values('Revenue', ascending=False).reset_index(drop=True)
sorted_by_revenue

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1016,Amit,North,Laptop,Electronics,6,500,3000
1,O1001,Asha,North,Laptop,Electronics,5,500,2500
2,O1023,Neha,South,Laptop,Electronics,4,500,2000
3,O1024,Amit,East,Monitor,Electronics,9,180,1620
4,O1024,Amit,East,Monitor,Electronics,9,180,1620
5,O1011,Neha,East,Laptop,Electronics,3,500,1500
6,O1004,Amit,East,Monitor,Electronics,7,180,1260
7,O1006,Ravi,West,Laptop,Electronics,2,500,1000
8,O1017,Asha,South,Monitor,Electronics,5,180,900
9,O1012,Amit,South,Printer,Office,4,220,880


## 3. Calculate Total Revenue

Use the `sum()` method.

In [6]:
total_revenue = df['Revenue'].sum()
total_revenue

np.int64(23240)

## 4. Calculate Average Revenue

Use the `mean()` method.

In [7]:
average_revenue = df['Revenue'].mean()
average_revenue

np.float64(929.6)

## 5. Count total records

`len(df)` counts rows, including rows with missing values in some columns.

In [8]:
total_records = len(df)
total_records

25

## 6. Count numeric Revenue cells

`count()` counts non-missing values in the Revenue column.

In [9]:
numeric_revenue_cells = df['Revenue'].count()
numeric_revenue_cells

np.int64(25)

## 7. COUNTIF: Orders from North region

Create a Boolean condition and sum `True` values.

In [10]:
north_orders = df['Region'].eq('North').sum()
north_orders

np.int64(6)

## 8. SUMIF: Total Revenue for Electronics

Filter the category and sum Revenue.

In [11]:
electronics_revenue = df.loc[df['Category'].eq('Electronics'), 'Revenue'].sum()
electronics_revenue

np.int64(16120)

## 9. AVERAGEIF: Average Units sold by Asha

Filter the salesperson and calculate the mean Units.

In [12]:
asha_average_units = df.loc[df['Salesperson'].eq('Asha'), 'Units'].mean()
asha_average_units

np.float64(8.333333333333334)

## 10. Highlight Revenue greater than 1,000

Use a pandas `Styler` to apply conditional formatting.

In [13]:
def highlight_high_revenue(value):
    return 'background-color: lightgreen; font-weight: bold' if value > 1000 else ''

df.style.map(highlight_high_revenue, subset=['Revenue']).format({'UnitPrice': '{:,.2f}', 'Revenue': '{:,.2f}'})

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500.00,"2,500.00"
1,O1002,Ravi,South,Mouse,Accessories,20,20.00,400.00
2,O1003,Neha,West,Keyboard,Accessories,15,30.00,450.00
3,O1004,Amit,East,Monitor,Electronics,7,180.00,"1,260.00"
4,O1005,Asha,North,Printer,Office,3,220.00,660.00
5,O1006,Ravi,West,Laptop,Electronics,2,500.00,"1,000.00"
6,O1007,Neha,South,Chair,Furniture,6,90.00,540.00
7,O1008,Amit,East,Desk,Furniture,2,200.00,400.00
8,O1009,Asha,West,Mouse,Accessories,25,20.00,500.00
9,O1010,Ravi,North,Monitor,Electronics,4,180.00,720.00


## 11. Apply a colour scale to Revenue

Use `background_gradient()` on the Revenue column.

In [14]:
df.style.background_gradient(subset=['Revenue'], cmap='RdYlGn').format({'UnitPrice': '{:,.2f}', 'Revenue': '{:,.2f}'})

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1001,Asha,North,Laptop,Electronics,5,500.00,"2,500.00"
1,O1002,Ravi,South,Mouse,Accessories,20,20.00,400.00
2,O1003,Neha,West,Keyboard,Accessories,15,30.00,450.00
3,O1004,Amit,East,Monitor,Electronics,7,180.00,"1,260.00"
4,O1005,Asha,North,Printer,Office,3,220.00,660.00
5,O1006,Ravi,West,Laptop,Electronics,2,500.00,"1,000.00"
6,O1007,Neha,South,Chair,Furniture,6,90.00,540.00
7,O1008,Amit,East,Desk,Furniture,2,200.00,400.00
8,O1009,Asha,West,Mouse,Accessories,25,20.00,500.00
9,O1010,Ravi,North,Monitor,Electronics,4,180.00,720.00


## 12. Identify duplicate OrderID values

Use `duplicated(keep=False)` to show every occurrence of a repeated ID.

In [15]:
duplicate_orders = df[df['OrderID'].duplicated(keep=False)]
duplicate_orders

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
23,O1024,Amit,East,Monitor,Electronics,9,180,1620
24,O1024,Amit,East,Monitor,Electronics,9,180,1620


## 13. Create a lookup table and perform a VLOOKUP-style lookup

In pandas, `map()` or `merge()` replaces VLOOKUP. First create a unique Product–Category lookup table.

In [16]:
product_lookup = df[['Product', 'Category']].drop_duplicates().sort_values('Product').reset_index(drop=True)
product_to_category = product_lookup.set_index('Product')['Category']
df_vlookup = df[['OrderID', 'Product']].copy()
df_vlookup['Category_from_lookup'] = df_vlookup['Product'].map(product_to_category)
product_lookup, df_vlookup.head(10)

(    Product     Category
 0     Chair    Furniture
 1      Desk    Furniture
 2  Keyboard  Accessories
 3    Laptop  Electronics
 4   Monitor  Electronics
 5     Mouse  Accessories
 6   Printer       Office,
   OrderID   Product Category_from_lookup
 0   O1001    Laptop          Electronics
 1   O1002     Mouse          Accessories
 2   O1003  Keyboard          Accessories
 3   O1004   Monitor          Electronics
 4   O1005   Printer               Office
 5   O1006    Laptop          Electronics
 6   O1007     Chair            Furniture
 7   O1008      Desk            Furniture
 8   O1009     Mouse          Accessories
 9   O1010   Monitor          Electronics)

## 14. Repeat using an XLOOKUP-style approach

A left `merge()` is the closest general equivalent of XLOOKUP.

In [17]:
df_xlookup = df[['OrderID', 'Product']].merge(product_lookup, on='Product', how='left', validate='many_to_one')
df_xlookup.head(10)

,OrderID,Product,Category
0,O1001,Laptop,Electronics
1,O1002,Mouse,Accessories
2,O1003,Keyboard,Accessories
3,O1004,Monitor,Electronics
4,O1005,Printer,Office
5,O1006,Laptop,Electronics
6,O1007,Chair,Furniture
7,O1008,Desk,Furniture
8,O1009,Mouse,Accessories
9,O1010,Monitor,Electronics


## 15. QUERY pivot: Sum of Revenue by Region

Use `groupby()` or `pivot_table()`.

In [18]:
revenue_by_region = df.groupby('Region', as_index=False)['Revenue'].sum().sort_values('Revenue', ascending=False)
revenue_pivot = df.pivot_table(index='Region', values='Revenue', aggfunc='sum').sort_values('Revenue', ascending=False)
revenue_by_region, revenue_pivot

(  Region  Revenue
 1  North     7580
 0   East     6960
 2  South     5140
 3   West     3560,
         Revenue
 Region         
 North      7580
 East       6960
 South      5140
 West       3560)

## 16. SORT formula: Display data by Units descending

Use `sort_values()`.

In [19]:
units_descending = df.sort_values(['Units', 'Revenue'], ascending=[False, False]).reset_index(drop=True)
units_descending

,OrderID,Salesperson,Region,Product,Category,Units,UnitPrice,Revenue
0,O1009,Asha,West,Mouse,Accessories,25,20,500
1,O1002,Ravi,South,Mouse,Accessories,20,20,400
2,O1019,Neha,East,Mouse,Accessories,18,20,360
3,O1003,Neha,West,Keyboard,Accessories,15,30,450
4,O1020,Amit,South,Keyboard,Accessories,14,30,420
5,O1013,Asha,North,Keyboard,Accessories,10,30,300
6,O1024,Amit,East,Monitor,Electronics,9,180,1620
7,O1024,Amit,East,Monitor,Electronics,9,180,1620
8,O1015,Neha,West,Chair,Furniture,8,90,720
9,O1004,Amit,East,Monitor,Electronics,7,180,1260


## 17. Find highest and lowest Revenue

Use `max()` and `min()`; `idxmax()` and `idxmin()` identify the corresponding rows.

In [20]:
highest_revenue = df['Revenue'].max()
lowest_revenue = df['Revenue'].min()
highest_row = df.loc[df['Revenue'].idxmax()]
lowest_row = df.loc[df['Revenue'].idxmin()]
{'Highest Revenue': highest_revenue, 'Lowest Revenue': lowest_revenue}, highest_row, lowest_row

({'Highest Revenue': 3000, 'Lowest Revenue': 200},
 OrderID              O1016
 Salesperson           Amit
 Region               North
 Product             Laptop
 Category       Electronics
 Units                    6
 UnitPrice              500
 Revenue               3000
 Name: 15, dtype: object,
 OrderID            O1014
 Salesperson         Ravi
 Region              East
 Product             Desk
 Category       Furniture
 Units                  1
 UnitPrice            200
 Revenue              200
 Name: 13, dtype: object)

## 18. Calculate each row's percentage contribution to total Revenue

Divide each row's Revenue by total Revenue and multiply by 100.

In [21]:
df_percentage = df.copy()
df_percentage['Revenue_Percentage'] = (df_percentage['Revenue'] / df_percentage['Revenue'].sum() * 100).round(2)
df_percentage[['OrderID', 'Revenue', 'Revenue_Percentage']]

,OrderID,Revenue,Revenue_Percentage
0,O1001,2500,10.76
1,O1002,400,1.72
2,O1003,450,1.94
3,O1004,1260,5.42
4,O1005,660,2.84
5,O1006,1000,4.30
6,O1007,540,2.32
7,O1008,400,1.72
8,O1009,500,2.15
9,O1010,720,3.10


## 19. Use IF logic to label Revenue as High or Low

Use `np.where()` for vectorised IF logic.

In [22]:
df_labelled = df.copy()
df_labelled['Revenue_Level'] = np.where(df_labelled['Revenue'] >= 1000, 'High', 'Low')
df_labelled[['OrderID', 'Revenue', 'Revenue_Level']]

,OrderID,Revenue,Revenue_Level
0,O1001,2500,High
1,O1002,400,Low
2,O1003,450,Low
3,O1004,1260,High
4,O1005,660,Low
5,O1006,1000,High
6,O1007,540,Low
7,O1008,400,Low
8,O1009,500,Low
9,O1010,720,Low


## 20. Protect formula cells — Python equivalent

A DataFrame has no editable formula cells. Protect the original data by keeping an untouched deep copy and creating derived columns in a separate result DataFrame.

In [23]:
df_original = df.copy(deep=True)
df_results = df_original.assign(
    Revenue_Percentage=lambda x: (x['Revenue'] / x['Revenue'].sum() * 100).round(2),
    Revenue_Level=lambda x: np.where(x['Revenue'] >= 1000, 'High', 'Low')
)
original_unchanged = df_original.equals(df)
original_unchanged, df_results.head()

(True,
   OrderID Salesperson Region   Product     Category  Units  UnitPrice  \
 0   O1001        Asha  North    Laptop  Electronics      5        500   
 1   O1002        Ravi  South     Mouse  Accessories     20         20   
 2   O1003        Neha   West  Keyboard  Accessories     15         30   
 3   O1004        Amit   East   Monitor  Electronics      7        180   
 4   O1005        Asha  North   Printer       Office      3        220   
 
    Revenue  Revenue_Percentage Revenue_Level  
 0     2500               10.76          High  
 1      400                1.72           Low  
 2      450                1.94           Low  
 3     1260                5.42          High  
 4      660                2.84           Low  )

## Consolidated answer summary

In [24]:
answer_summary = pd.DataFrame({
    'Measure': [
        'Total Revenue', 'Average Revenue', 'Total Records',
        'Numeric Revenue Cells', 'North Region Orders',
        'Electronics Revenue', 'Average Units sold by Asha',
        'Highest Revenue', 'Lowest Revenue'
    ],
    'Answer': [
        total_revenue, average_revenue, total_records,
        numeric_revenue_cells, north_orders,
        electronics_revenue, asha_average_units,
        highest_revenue, lowest_revenue
    ]
})
answer_summary

,Measure,Answer
0,Total Revenue,"23,240.00"
1,Average Revenue,929.60
2,Total Records,25.00
3,Numeric Revenue Cells,25.00
4,North Region Orders,6.00
5,Electronics Revenue,"16,120.00"
6,Average Units sold by Asha,8.33
7,Highest Revenue,"3,000.00"
8,Lowest Revenue,200.00


## Additional useful analysis

In [25]:
salesperson_summary = df.groupby('Salesperson', as_index=False).agg(
    Orders=('OrderID', 'count'),
    Units=('Units', 'sum'),
    Revenue=('Revenue', 'sum'),
    Average_Revenue=('Revenue', 'mean')
).sort_values('Revenue', ascending=False)

category_summary = df.groupby('Category', as_index=False).agg(
    Orders=('OrderID', 'count'),
    Units=('Units', 'sum'),
    Revenue=('Revenue', 'sum')
).sort_values('Revenue', ascending=False)

salesperson_summary, category_summary

(  Salesperson  Orders  Units  Revenue  Average_Revenue
 0        Amit       7     51     9200         1,314.29
 2        Neha       6     54     5570           928.33
 1        Asha       6     50     5260           876.67
 3        Ravi       6     34     3210           535.00,
       Category  Orders  Units  Revenue
 1  Electronics      10     54    16120
 2    Furniture       6     24     2710
 0  Accessories       6    102     2430
 3       Office       3      9     1980)

## Key observations

- Total Revenue is **23,240** and Average Revenue is **929.60**.
- The North region has the highest regional Revenue at **7,580**.
- Electronics contributes **16,120**, the largest category total.
- The highest single-row Revenue is **3,000**; the lowest is **200**.
- `O1024` appears twice and should be reviewed before final reporting because it may duplicate Revenue by **1,620**.
